### PR TITLE
Kruskal Algorithm in C

### LINKED ISSUE #1991: Kruskal Algorithm in C
@jainaman224 I would like to work for kruskal algorithm in C. Please assign this issue to me for GSSoC.

### PR DESCRIPTION
**Pull request by coderprasukj**
Fixes #1991 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Kruskal Algorithm in C

#### Languages Used:
- C

#### Files Added:
- Kruskal_Algorithm.c

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 1]
```diff
@@ -0,0 +1,133 @@
+// Kruskal Algorithm in C
+#include <stdio.h>
+#define NUM 20
+
+// Creating the edge
+ typedef struct edge {
+	int p;          // Variable to contain all the first node of the edge
```
Comment (jainaman224) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Use 4 space indentation.

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	int q;          // Variable to contain all the second node of the edge  
+	int weight;     // Variable to contain the weight of the edge
+}edge;
```
Comment (jainaman224) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add a space after `}`.

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+ 
+// Creating the Edgelist
+typedef struct edgelist {
+	edge data[NUM];
+	int n;
+}edgelist;
+ 
+edgelist list;
+int G[NUM][NUM];          // Array for weights in the graph
```
Comment (jainaman224) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Use better variable name than G.

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	
+	for(int i=0; i<n; i++)                       // Enter the Adjacency Matrix
+		for(int j=0; j<n; j++)
+			scanf("%d",&G[i][j]);
+			
+	kruskal_algo();
+	print();
+	return 0;
+}
+// Function performing Kruskal's algorithm
```
Comment (jainaman224) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add a new line above this.

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	printf("\nCost of the spanning tree = %d\n",cost);
+}
+
+
+/*
+INPUT :
+Enter number of vertices : 6
+Enter the Adjacency Matrix :
+		0  3  0  0  6  5
```
Comment (jainaman224) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Align this.

**Comment by coderprasukj** [comment]
@jainaman224 Sir I have done the required changes. Please go through it.

File: Kruskal_Algorithm/Kruskal_Algorithm.c (hunk 1 above)
Comment (jainaman224) [reply Kruskal_Algorithm/Kruskal_Algorithm.c]: Not done.

**Comment by manildayma** [comment]
Can I do this in python, C#?

**Comment by pragyajaiswa05l** [comment]
please assign this issue to me would like to implement it in c++ @jainaman224

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+void kruskal_algo();
+void print();
+int find(int contain[], int vertexno);
+void union1(int contain[], int c1, int c2);
+void sort();
+
+
+// Driver Program
+int main() {
+	int total_cost;
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Please use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	print();
+	return 0;
+}
+
+// Function performing Kruskal's algorithm
+void kruskal_algo() {
+	int contain[NUM], x, y;
+	list.n=0;
+ 
+	for(int i=1; i<n; i++)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Please use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+		if(x != y) {
+			listSpan.data[listSpan.n] = list.data[i];
+			listSpan.n = listSpan.n+1;
+			union1(contain, x, y);
+		}
+	}
+}
+ 
+// Returns the value present at k[vertex_number]
+int find(int contain[],int vertex_number) {
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Please use 4 space indentation

**Comment by somya-kapoor** [comment]
@coderprasukj Please use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c (hunk 6 above)
Comment (coderprasukj) [reply Kruskal_Algorithm/Kruskal_Algorithm.c]: @somya-kapoor I am not knowing about 4 space indentation. Can you please tell me one time by an example.

**Comment by coderprasukj** [comment]
@somya-kapoor I have modified the changes in my file. Please review it.

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+typedef struct edge {
+	int p;          // Variable to contain all the first node of the edge
+	int q;          // Variable to contain all the second node of the edge  
+	int weight;     // Variable to contain the weight of the edge
+} edge;
+ 
+// Creating the Edgelist
+typedef struct edgelist {
+    edge data[NUM];
+	int n;
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: inconsistency in indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+// Driver Program
+int main() {
+	int total_cost; 
+	printf("\nEnter number of vertices:");
+	scanf("%d",&n);                              // Number of Vertices
+	
+	printf("\nEnter the adjacency matrix:\n");
+	
+	for(int i=0; i<n; i++)                       // Enter the Adjacency Matrix
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space around the operator

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+		for(int j=0; j<n; j++)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space around the operator

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+			scanf("%d", &weights[i][j]);
+			
+	kruskal_algo();
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+	print();
+	return 0;
+}
+
+// Function performing Kruskal's algorithm
+void kruskal_algo() {
+	int contain[NUM], x, y;
+	list.n = 0;
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+ 
+	for(int i=1; i<n; i++)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space around operator

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+		for(int j=0; j<i; j++) {
+			if(weights[i][j] != 0) {
+				list.data[list.n].p = i;          // Containing all the first node of the edges
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+				list.data[list.n].q = j;          // Containing all the second node of the edges
+				list.data[list.n].weight = weights[i][j];   // Containing the weight of the edges
+				list.n++;
+			}
+		}
+
+	// Sorting the edges according to their weight.
+	sort();
+	
+	for(int i=0; i<n; i++)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space around operator

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+		contain[i] = i;            // Intializing the array with first node
+	
+	listSpan.n = 0;
+	
+	for(int i=0; i<list.n; i++) {
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space around operator

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+	}
+}
+ 
+// Returns the value present at k[vertex_number]
+int find(int contain[], int vertex_number) {
+	return(contain[vertex_number]);
+}
+ 
+void union1(int contain[],int c1,int c2) {
+	for(int i=0;i<n;i++)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space around operator

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+		if(contain[i]==c2)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space around operator

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+			contain[i]=c1;
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space around operator

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+}
+ 
+// Sorting all the edges in increasing order
+// according to their weight.
+void sort() {
+	edge temp;
+	for(int i=1; i<list.n; i++)
+		for(int j=0; j<list.n-1; j++)
+			if(list.data[j].weight > list.data[j+1].weight) {
+				temp=list.data[j];
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: use 4 space indentation

**Comment by coderprasukj** [comment]
@somya-kapoor I have modified the file according to changes required. Please Review it.

**Comment by somya-kapoor** [comment]
@coderprasukj Please use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+{
+	int cost = 0;
+	for(int i = 0; i < listSpan.n; i++) 
+	{
+		printf("\n%d\t%d\t%d", listSpan.data[i].p, listSpan.data[i].q, listSpan.data[i].weight);
+		cost = cost + listSpan.data[i].weight;
+	}
+	printf("\nCost of the spanning tree = %d\n",cost);
+}
+
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Remove this line

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+// Printing the Output
+void print() 
+{
+	int cost = 0;
+	for(int i = 0; i < listSpan.n; i++) 
+	{
+		printf("\n%d\t%d\t%d", listSpan.data[i].p, listSpan.data[i].q, listSpan.data[i].weight);
+		cost = cost + listSpan.data[i].weight;
+	}
+	printf("\nCost of the spanning tree = %d\n",cost);
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add space after ,

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+ 
+// Returns the value present at k[vertex_number]
+int find(int contain[], int vertex_number) 
+{
+	return(contain[vertex_number]);
+}
+ 
+void union1(int contain[], int c1, int c2) 
+{
+	for(int i = 0; i < n;i++)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: add space after ;

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+void union1(int contain[], int c1, int c2);
+void sort();
+
+
+// Driver Program
+int main() 
+{
+	int total_cost; 
+	printf("\nEnter number of vertices : ");
+	scanf("%d",&n);                              // Number of Vertices
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c]: Add a space after ,

**Comment by somya-kapoor** [comment]
@coderprasukj Are you working on it?

**Comment by coderprasukj** [comment]
@somya-kapoor I have done the required changes. Please go through it.

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 26]
```diff
@@ -0,0 +1,152 @@
+// Kruskal Algorithm in C
+#include <stdio.h>
+#define NUM 20
+
+// Creating the edge
+typedef struct edge 
+{
+	int p;          // Variable to contain all the first node of the edge
+	int q;          // Variable to contain all the second node of the edge
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:9]: You are using 8 space indentation. Use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+{
+	int p;          // Variable to contain all the first node of the edge
+	int q;          // Variable to contain all the second node of the edge  
+	int weight;     // Variable to contain the weight of the edge
+} edge;
+ 
+// Creating the Edgelist
+typedef struct edgelist 
+{
+	edge data[NUM];
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:16]: You are using 8 space indentation. Use 4 space indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+void print();
+int find(int contain[], int vertexno);
+void union1(int contain[], int c1, int c2);
+void sort();
+
+
+// Driver Program
+int main() 
+{
+	int total_cost;
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:35]: You are using 8 space indentation. Use 4 space indentation

**Comment by somya-kapoor** [comment]
You are using 8 space indentation. Use 4 space indentation

**Comment by coderprasukj** [comment]
@somya-kapoor Done with the changes. Please go through it once.

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+int weights[NUM][NUM];    // Array for weights in the graph
+int n;
+edgelist listSpan;
+ 
+void kruskal_algo();
+void print();
+int find(int contain[], int vertexno);
+void union1(int contain[], int c1, int c2);
+void sort();
+
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:30]: Remove this line

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+
+// Driver Program
+int main() 
+{
+    int total_cost; 
+    printf("\nEnter number of vertices : ");
+	scanf("%d", &n);    // Number of Vertices
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:37]: inconsistency in indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+	
+    printf("\nEnter the adjacency matrix : \n");
+	
+    for(int i = 0; i < n; i++)    // Enter the Adjacency Matrix
+	    for(int j = 0; j < n; j++)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:42]: inconsistency in indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+// Function performing Kruskal's algorithm
+void kruskal_algo() 
+{
+    int contain[NUM], x, y;
+    list.n = 0;
+ 
+    for(int i = 1; i < n; i++)
+	    for(int j = 0; j < i; j++)
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:57]: inconsistency in indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 33]
```diff
@@ -0,0 +1,151 @@
...
+ 
+    for(int i = 1; i < n; i++)
+	    for(int j = 0; j < i; j++) 
+        {
+            if(weights[i][j] != 0) 
+            {
+                list.data[list.n].p = i;    // Containing all the first node of the edges
+                list.data[list.n].q = j;    // Containing all the second node of the edges
+                list.data[list.n].weight = weights[i][j];    // Containing the weight of the edges
+                list.n++;
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:64]: inconsistency in indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 34]
```diff
@@ -0,0 +1,151 @@
...
+            }
+       }
+
+    // Sorting the edges according to their weight.
+    sort();
+	
+    for(int i = 0; i < n; i++)
+	    contain[i] = i;    // Intializing the array with first node
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:72]: inconsistency in indentation

File: Kruskal_Algorithm/Kruskal_Algorithm.c [hunk 35]
```diff
@@ -0,0 +1,151 @@
...
+{
+    int cost = 0;
+    for(int i = 0; i < listSpan.n; i++) 
+    {
+        printf("\n%d\t%d\t%d", listSpan.data[i].p, listSpan.data[i].q, listSpan.data[i].weight);
+        cost = cost + listSpan.data[i].weight;
+    }
+    printf("\nCost of the spanning tree = %d\n", cost);
+}
+/*
```
Comment (somya-kapoor) [inline Kruskal_Algorithm/Kruskal_Algorithm.c:132]: leave a line before it

**Comment by coderprasukj** [comment]
@somya-kapoor I have done the required changes in latest commit. Please have a go through to it.

**Comment by coderprasukj** [comment]
@somya-kapoor I have sqauashed the commits. Please have a look at it.

**Pull request merged**